core/vote/vote_pool.go

```diff
@@ -0,0 +1,401 @@
+const (
+	maxCurVoteAmountPerBlock    = 21
+	maxFutureVoteAmountPerBlock = 50
+
+	voteBufferForPut            = 256
+	lowerLimitOfVoteBlockNumber = 256
+	upperLimitOfVoteBlockNumber = 13
+
+	chainHeadChanSize = 10 // chainHeadChanSize is the size of channel listening to ChainHeadEvent.
+)
+
+var (
+	localCurVotesGauge    = metrics.NewRegisteredGauge("curVotes/local", nil)
+	localFutureVotesGauge = metrics.NewRegisteredGauge("futureVotes/local", nil)
+
+	localReceivedVotesGauge = metrics.NewRegisteredGauge("receivedVotes/local", nil)
+
+	localCurVotesPqGauge    = metrics.NewRegisteredGauge("curVotesPq/local", nil)
+	localFutureVotesPqGauge = metrics.NewRegisteredGauge("futureVotesPq/local", nil)
+)
+
+type VoteBox struct {
+	blockNumber  uint64
+	voteMessages []*types.VoteEnvelope
+}
+
+type VotePool struct {
+	chain       *core.BlockChain
+	chainconfig *params.ChainConfig
+	mu          sync.RWMutex
+
+	votesFeed event.Feed
+	scope     event.SubscriptionScope
+
+	receivedVotes mapset.Set
+
+	curVotes    map[common.Hash]*VoteBox
+	futureVotes map[common.Hash]*VoteBox
+
+	curVotesPq    *votesPriorityQueue
+	futureVotesPq *votesPriorityQueue
+
+	chainHeadCh  chan core.ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	votesCh chan *types.VoteEnvelope
+
+	engine consensus.PoSA
+}
+
+type votesPriorityQueue []*types.VoteData
+
+func NewVotePool(chainconfig *params.ChainConfig, chain *core.BlockChain, engine consensus.PoSA) *VotePool {
+	votePool := &VotePool{
+		chain:         chain,
+		chainconfig:   chainconfig,
+		receivedVotes: mapset.NewSet(),
+		curVotes:      make(map[common.Hash]*VoteBox),
+		futureVotes:   make(map[common.Hash]*VoteBox),
+		curVotesPq:    &votesPriorityQueue{},
+		futureVotesPq: &votesPriorityQueue{},
+		chainHeadCh:   make(chan core.ChainHeadEvent, chainHeadChanSize),
+		votesCh:       make(chan *types.VoteEnvelope, voteBufferForPut),
+		engine:        engine,
+	}
+
+	// Subscribe events from blockchain and start the main event loop.
+	votePool.chainHeadSub = votePool.chain.SubscribeChainHeadEvent(votePool.chainHeadCh)
+	if posa, ok := engine.(consensus.PoSA); ok {
+		posa.SetVotePool(votePool)
+	}
+
+	go votePool.loop()
+	return votePool
+}
+
+// loop is the vote pool's main even loop, waiting for and reacting to outside blockchain events and votes channel event.
+func (pool *VotePool) loop() {
+	for {
+		select {
+		// Handle ChainHeadEvent.
+		case ev := <-pool.chainHeadCh:
+			if ev.Block != nil {
+				latestBlockNumber := ev.Block.NumberU64()
+				pool.prune(latestBlockNumber)
+				pool.transferVotesFromFutureToCur(ev.Block.Header())
+			}
+		case <-pool.chainHeadSub.Err():
+			return
+
+		// Handle votes channel and put the vote into vote pool.
+		case vote := <-pool.votesCh:
+			pool.putIntoVotePool(vote)
+		}
+	}
+}
+
+func (pool *VotePool) PutVote(vote *types.VoteEnvelope) {
+	pool.votesCh <- vote
+}
+
+func (pool *VotePool) putIntoVotePool(vote *types.VoteEnvelope) bool {
+	targetNumber := vote.Data.TargetNumber
```

Comment:
if this vote is too stale, please discard it immediately.

Reply:
Updated, will check height range firstly